src/wallet/wallet.cpp

```diff
@@ -100,7 +100,7 @@ CPubKey CWallet::GenerateNewKey()
 
     // use HD key derivation if HD was enabled during wallet creation
     if (IsHDEnabled()) {
-        DeriveNewChildKey(metadata, secret);
+        DeriveNewChildKey(metadata, secret, (CanSupportFeature(FEATURE_HD_SPLIT) ? internal : false));
```

Comment:
A design question, but IMO `DeriveNewChildKey` should execute the derivation and not care about wallet versions and supported features.